### ISSUE TITLE
Collapse not working

### ISSUE THREAD

**Issue by GrmpCerber** [amaembo/streamex#29]
I didn't found where it came from but the following code does not work:

    streamEx
        .sorted()
        .collapse(C::isNested,C::merge)
        .collect(toList());
        ...
        //items are not collapsed
        
        public static final boolean isNested(T a, T b) {
            return a.isParent(b) || b.isParent(a);
        };

        public static final T merge(T a, T b) {
            return a.isParent(b) ? a : b;
        };

"POJO" is working :

    List<T> tmp = 
        stream
            .sorted()
            .collect(toList());
    Iterator<T> it = tmp.iterator();
    T curr, last;
    curr = last = null;
    while (it.hasNext()) {
        T oldLast = last;
        last = curr;
        curr = it.next();
        if (last != null && last.isParent(curr)) {
            it.remove();
            curr = last;
            last = oldLast;
        }
    }
    tmp.stream().collect(toList());
    //items are sorted

**Comment by amaembo** [comment]
Thank you for the report. Unfortunately I cannot reproduce your problem. See the [unit test](https://github.com/amaembo/streamex/blob/cbbfd4b7c32beba86996e83ac7f486aa75d156b2/src/test/java/one/util/streamex/issues/Issue0029.java). Please provide the complete failing test (or edit the one I committed).

Note that you are doing the different things: in "POJO" approach you collapse only if `last.isParent(curr)`, but in "streamEx" approach you collapse also if `curr.isParent(last)`.

Probably the origin of your problem is the fact that `BiPredicate` of [`collapse`](http://amaembo.github.io/streamex/javadoc/one/util/streamex/StreamEx.html#collapse-java.util.function.BiPredicate-java.util.function.BinaryOperator-) method family always receives the *element from original stream* (not the result of the previous merge). Anyways, provide the full test: probably your problem still can be solved with collapse.

**Comment by GrmpCerber** [comment]
Here's a full working test.
It's true that my StreamEx version was coded slightly differently. I did it on purpose since I was unsure of the element's order.

**Comment by amaembo** [comment]
Thank you, I see now what's the problem. The `collapse()` method works as expected: the first time it passes `192.168.0.0/16` and `192.168.0.0/24` to the `BiPredicate`, it returns true and actually merged (you can check that result contains no `192.168.0.0/24`). But the second time it passes `192.168.0.0/24` and `192.168.1.0/24` to `BiPredicate` and it returns false.

As I said before, `collapse()` calls `BiPredicate` with two adjacent elements of the *original stream*, not the result of the previous merge. One of the reason for such design is to support parallel processing. Suppose that your input is split to two halves which are processed by separate threads. In this case the second thread will have `192.168.25.0/24` and so on up to `192.168.49.0/24`. None of these elements are parents to each other, so they cannot be merged at all (at least until the first half is processed).

Nevertheless your problem is interesting. I will think how to solve it easier with StreamEx or plain Stream API. Probably custom collector could be implemented which does the thing. Not closing the issue for now.

**Comment by amaembo** [comment]
I think up some collector-based solution and updated [your example](https://github.com/amaembo/streamex/blob/421df5df8d2e46dc049011f7aa8bfcfd06a026ee/src/test/java/one/util/streamex/issues/Issue0029.java). See `testCollect()` test which passes now for both sequential and parallel stream without preliminary sorting. You need `getLowerKey()` and `getHidherKey()`, but don't need anymore `isParent()` and `compareTo()` methods.

The idea is to create the collector for interval objects (which IP ranges are) using user-provided functions to extract the comparable "lowerKey" and "higherKey" objects. The only requirement is that input intervals cannot partially overlap: either A contains B totally or B contains A totally or they don't overlap at all. CIDR entries fulfill this requirement. This collector should correctly merge input intervals if one of them covers another one. Such collector could be added to my `MoreCollectors` class as it's quite generic and does not depend on particular problem of collapsing the IP ranges. What do you think?

**Comment by GrmpCerber** [comment]
Got it. It should work in my use case.
Yet, just to mention it, this is not a perfect solution since we loose the <em>quasi-intermediate</em> characteristic of your collapse operation.
Do you think this could be solve with a <em>quasi-intermediate</em> operation ?

**Comment by amaembo** [comment]
Without sorting it would be full-stop intermediate operation: downstream processing cannot begin until all the upstream processing is done. Now the only such operation in Stream API is `sorted()`. I don't want to add new full-stop intermediate operations (except clearly derived from `sorted()` as `sortedBy()`) as it's quite painful to create parallel implementation and also such operation has less predictable performance/memory usage characteristics.

It's probably possible to implement it assuming that the input is pre-sorted (as in your attempt). However note that sorting is unnecessary step here. It takes additional O(n) memory (to store all elements into the intermediate array/list and sort it), but online processing via TreeMap may take much less memory (for your input data as they are already sorted you never have more than two items in the TreeMap). Thus I'd prefer terminal solution (Collector). This way it would be more clear to the user: this operation creates hard-copy of your data. If you want to continue streaming, you may create new stream from that list.

I rethink my solution. Probably it's still better to define your problem with `Comparator` and `BiPredicate` (which establishes parent-child relation) than by extracting lowerKey and higherKey. There's similar problem to collapse set of paths like `/a/b/*`, `/a/c/*`, `/a/b/c/*`, `/a/b/c/d/*` (we want to remove the latter two paths as they are superceded with the first path). In this case providing higherKey would be quite artificial, while `Comparator` and `BiPredicate` look more natural. So probably I will provide another version.

**Comment by amaembo** [comment]
I committed the updated [solution](https://github.com/amaembo/streamex/blob/c5c51481211eaebe7e55899c0a71390108da540d/src/test/java/one/util/streamex/issues/Issue0029.java). Also I simplified your `compareTo` method. You may consider checking it on bigger input dataset.

**Comment by GrmpCerber** [comment]
Thanks,
As I said it was a "simplified (and ugly+hardcoded) version of my work" ;)

**Comment by amaembo** [comment]
I'm closing this issue as the original report "Collapse not working" is invalid (`collapse()` actually works as documented). However I opened a new Issue #32 to introduce a new `Collector` which will solve your problem. Thank you again for your report.

**Comment by amaembo** [comment]
Finally I came to the solution which I like (see #32). My idea of using `TreeMap` actually rarely outperforms your approach on pre-sorting the input, so I created a `dominators()` collector instead which requires (for your problem) that the input is already sorted. Such collector is more generic as it can solve some interesting problems even with non-sorted inputs. So your problem will be solved like this:

    StreamEx.of(getTestData()).sorted().collect(MoreCollectors.dominators(T::isParent));

As I mentioned before, I don't like making such operation as quasi-intermediate, because it's impossible to provide efficient parallel implementation: I cannot even start sending the second part elements to the downstream without fully processing the first part. But having it as collector solves the problem perfectly as I can remove some elements from the resulting list if necessary when the first part arrives.

Thank you again for the interesting problem.
